File: battle/attack/AttackWave.java
package common.battle.attack;

import common.battle.entity.AbEntity;
import common.battle.entity.Entity;
import common.util.Data;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class AttackWave extends AttackAb {

	protected final Set<Entity> incl;

	public AttackWave(Entity e, AttackSimple a, double p0, double wid, int wt) {
		super(e, a, p0 - wid / 2, p0 + wid / 2, false);
		waveType = wt;
		isCounter = a.isCounter;
		incl = new HashSet<>();
		if(waveType == WT_MINI)
			proc.MINIWAVE.lv--;
		else
			proc.WAVE.lv--;
	}

	public AttackWave(Entity e, AttackWave a, double p0, double wid) {
		super(e, a, p0 - wid / 2, p0 + wid / 2, false);
		waveType = a.waveType;
		isCounter = a.isCounter;
		incl = a.incl;
		if(waveType == WT_MINI)
			proc.MINIWAVE.lv--;
		else
			proc.WAVE.lv--;
	}

	public AttackWave(Entity e, AttackWave a, double pos, double start, double end) {
		super(e, a, pos - start, pos + end, false);
		waveType = a.waveType;
		isCounter = a.isCounter;
		incl = a.incl;
		if(waveType == WT_MINI)
			proc.MINIWAVE.lv--;
		else
			proc.WAVE.lv--;
	}

	@Override
	public void capture() {
		List<AbEntity> le = model.b.inRange(touch, dire, sta, end, excludeLastEdge);
		if (incl != null)
			le.removeIf(incl::contains);
		capt.clear();
		if ((abi & AB_ONLY) == 0)
			capt.addAll(le);
		else
			for (AbEntity e : le)
				if (e.ctargetable(trait, attacker, true))
					capt.add(e);
	}

	@Override
	public void excuse() {
		process();

		if(attacker != null) {
			if(attacker.status[P_WEAK][0] != 0) {
				atk = rawAtk * attacker.status[P_WEAK][1] / 100;
			}

			if(attacker.status[P_STRONG][0] != 0) {
				atk += rawAtk * attacker.status[P_STRONG][0] / 100;
			}

			if (attacker.dire == -1 && attacker.basis.canon.deco == DECO_BASE_WALL) {
				atk = (int) (atk * attacker.basis.b.t().getDecorationMagnification(attacker.basis.canon.deco, Data.DECO_WAVE));
			}
		}

		for (AbEntity e : capt) {
			if (e.isBase())
				continue;

			if (e instanceof Entity) {
				e.damaged(this);
				incl.add((Entity) e);
			}
		}
	}

}

File: battle/attack/AttackWave.kt
package common.battle.attack

import common.battle.entity.AbEntity
import common.battle.entity.Entity
import common.util.Data
import java.util.*
import java.util.function.Predicate

class AttackWave : AttackAb {
    val incl: MutableSet<Entity>?

    constructor(a: AttackSimple, p0: Double, wid: Double, wt: Int) : super(a, p0 - wid / 2, p0 + wid / 2) {
        waveType = wt
        incl = HashSet()
        proc.WAVE.lv--
    }

    constructor(a: AttackWave, p0: Double, wid: Double) : super(a, p0 - wid / 2, p0 + wid / 2) {
        waveType = a.waveType
        incl = a.incl
        proc.WAVE.lv--
    }

    override fun capture() {
        val le: MutableList<AbEntity> = model.b.inRange(touch, dire, sta, end)
        if (incl != null) le.removeIf(Predicate<AbEntity> { e: AbEntity -> incl.contains(e) })
        capt.clear()
        if (abi and Data.Companion.AB_ONLY == 0) capt.addAll(le) else for (e in le) if (e.targetable(type)) capt.add(e)
    }

    override fun excuse() {
        process()
        for (e in capt) {
            if (e.isBase()) continue
            if (e is Entity) {
                e.damaged(this)
                incl!!.add(e)
            }
        }
    }
}
